### PR TITLE
Option Redesign (fix for #513)

### LINKED ISSUE #513: Slash command option types are broken by stringized annotations
### Summary

When specifying slash command option types in a module that uses `from __future__ import annotations`, which stringizes annotations, a crash occurs.

### Reproduction Steps

~First of all, the current examples for the slash commands doesn't tell you to use type annotations for slash command option parameters, which is required to make them work.~ After adding an annotation to create an `int`-type option, my bot sadly crashed because it called `issubclass('int', str)` in `SlashCommandOptionType.from_datatype` in `discord/enums.py.` Clearly, that function expected `int` not `'int'`. I believe that the `SlashCommandOptionType.from_datatype` method is receiving parameters obtained by the python `inspect.signature(callable)` function, which unstringizes parameters from modules that use `from __future__ import annotations` only if its `eval_str` keyword argument is explicitly set to `True`, which it isn't currently. EDIT: actually, the `eval_str` keyword argument appears to be only available in Python 3.10 and not Python 3.9, which might necessitate a manual workaround for earlier versions. Also, re-reading the examples, it was my mistake to use a raw `int` type annotation instead of an `Option` type annotation in the first place, although it Almost worked; not sure if that is intended to be supported or not. Anyway, the problem still exists if you use Option type annotations properly.

### Minimal Reproducible Code

```
from __future__ import annotations
import discord
from discord.commands import Option

bot = discord.Bot()
@bot.slash_command()
async def hello(ctx, number: Option(int)):
    """Say hello to the bot"""
    await ctx.respond(f"Hello number {number}!")

bot.run("TOKEN")
```

### Expected Results

I expected to make a slash command with a typed option.

### Actual Results

Crashed. Full traceback:
```
Traceback (most recent call last):
  File "C:\Users\Extreme PC\.virtualenvs\mitchbot-YqVhPTxY\lib\site-packages\discord\client.py", line 352, in _run_event
    await coro(*args, **kwargs)
  File "C:\Users\Extreme PC\Code\mitchbot\MitchBot.py", line 37, in on_ready
    add_responses(self)
  File "C:\Users\Extreme PC\Code\mitchbot\responders.py", line 193, in add_responses
    async def get_nicknames(
  File "C:\Users\Extreme PC\.virtualenvs\mitchbot-YqVhPTxY\lib\site-packages\discord\bot.py", line 512, in decorator
    result = command(**kwargs)(func)
  File "C:\Users\Extreme PC\.virtualenvs\mitchbot-YqVhPTxY\lib\site-packages\discord\commands\commands.py", line 1110, in decorator
    return cls(func, **attrs)
  File "C:\Users\Extreme PC\.virtualenvs\mitchbot-YqVhPTxY\lib\site-packages\discord\commands\commands.py", line 393, in __init__
    self.options: List[Option] = kwargs.get('options') or self._parse_options(params)
  File "C:\Users\Extreme PC\.virtualenvs\mitchbot-YqVhPTxY\lib\site-packages\discord\commands\commands.py", line 449, in _parse_options
    option = Option(option, "No description provided")
  File "C:\Users\Extreme PC\.virtualenvs\mitchbot-YqVhPTxY\lib\site-packages\discord\commands\commands.py", line 618, in __init__
    _type = SlashCommandOptionType.from_datatype(input_type)
  File "C:\Users\Extreme PC\.virtualenvs\mitchbot-YqVhPTxY\lib\site-packages\discord\enums.py", line 643, in from_datatype
    if issubclass(datatype, str):
TypeError: issubclass() arg 1 must be a class
```

### Intents

I am not passing a discord.Intents class to my client.

### System Information

- Python v3.9.1-final
- py-cord v2.0.0-alpha
    - py-cord pkg_resources: v2.0.0a4477+g57ed00a0
- aiohttp v3.8.1
- system info: Windows 10 10.0.19041

### Checklist

- [X] I have searched the open issues for duplicates.
- [X] I have shown the entire traceback, if possible.
- [X] I have removed my token from display, if visible.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by EmreTech**
## Summary

This PR fixes #513 by redesigning how Options are set. This also documents the Option class.

TODO List for this PR to be merged:
- [x] Update the `option` decorator to reflect how Options are set now
- [x] Fix options defined in the slash_command kwarg
- [x] Document Option

## Checklist

- [x] If code changes were made then they have been tested.
    - [ ] I have updated the documentation to reflect the changes.
- [ ] If `type: ignore` comments were used, a comment is also left explaining why
- [x] This PR fixes an issue.
- [ ] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, typehinting, examples, ...)

### CODE REVIEW

File: discord/commands/options.py
Comment (VincentRPS) [inline discord/commands/options.py:28]: might be a bit dumb, but how are these used?

Comment (EmreTech) [reply discord/commands/options.py:28]: Well, because they are included, it means they would be stored in the globals() function, and thus the eval function below would recognize most legal module types

**Comment by Lulalaby** [comment]
@EmreTech what's the current state of this

**Comment by EmreTech** [comment]
> @EmreTech what's the current state of this

I've tested it, and it should be done now. I'm just waiting for reviews.

**Comment by Middledot** [comment]
This seems to break if parentheses are in the option description.
```python
@client.slash_command()
async def dis_is_test(ctx, channel: discord.Option(discord.TextChannel, description="Is a channel? (yes)")):
    await ctx.respond(channel.mention)
```
Value that is passed into `gettype`: `TextChannel, description='Is a channel? (yes)'` instead of only `TextChannel`

**Comment by EmreTech** [comment]
> This seems to break if parentheses are in the option description.
> 
> ```python
> @client.slash_command()
> async def dis_is_test(ctx, channel: discord.Option(discord.TextChannel, description="Is a channel? (yes)")):
>     await ctx.respond(channel.mention)
> ```
> 
> Value that is passed into `gettype`: `TextChannel, description='Is a channel? (yes)'` instead of only `TextChannel`

Working on it, thanks for telling me

**Comment by EmreTech** [comment]
> This seems to break if parentheses are in the option description.
> 
> ```python
> @client.slash_command()
> async def dis_is_test(ctx, channel: discord.Option(discord.TextChannel, description="Is a channel? (yes)")):
>     await ctx.respond(channel.mention)
> ```
> 
> Value that is passed into `gettype`: `TextChannel, description='Is a channel? (yes)'` instead of only `TextChannel`

Done.

**Review by Lulalaby — APPROVED** [review]

**Comment by toBeOfUse** [comment]
Just so you know, this fix seems to suffer from the problem I warned about in the comments of the original issue, where if you import the Option class and give it a name other than Option, the code can fail. For example:

```python3
from __future__ import annotations
import discord
from discord.commands import Option as OptionForSlashCommand

bot = discord.Bot()
@bot.slash_command()
async def hello(ctx, number: OptionForSlashCommand(int)):
    """Say hello to the bot"""
    await ctx.respond(f"Hello number {number}!")

bot.run("TOKEN")
```

will still crash immediately. I think this is a potential problem especially since "Option" is a very common class name that could easily conflict with some other code or dependencies that someone has, or maybe they just want to give it a name that makes it clear to them what it is for.

I do not know enough about Python type annotations to say exactly how to solve this, but @Angelin01 had a proposed solution in #513 involving using types as type annotations rather than, as is currently the case, objects as type annotations, which makes sense to me.

**Comment by EmreTech** [comment]
> Just so you know, this fix seems to suffer from the problem I warned about in the comments of the original issue, where if you import the Option class and give it a name other than Option, the code can fail. For example:
> 
> 
> 
> ```python3
> 
> from __future__ import annotations
> 
> import discord
> 
> from discord.commands import Option as OptionForSlashCommand
> 
> 
> 
> bot = discord.Bot()
> 
> @bot.slash_command()
> 
> async def hello(ctx, number: OptionForSlashCommand(int)):
> 
>     """Say hello to the bot"""
> 
>     await ctx.respond(f"Hello number {number}!")
> 
> 
> 
> bot.run("TOKEN")
> 
> ```
> 
> 
> 
> will still crash immediately. I think this is a potential problem especially since "Option" is a very common class name that could easily conflict with some other code or dependencies that someone has, or maybe they just want to give it a name that makes it clear to them what it is for.
> 
> 
> 
> I do not know enough about Python type annotations to say exactly how to solve this, but @Angelin01 had a proposed solution in #513 involving using types as type annotations rather than, as is currently the case, objects as type annotations, which makes sense to me.

I am very aware of what Angelin01 proposed (I used solution 1 as a basis for this entire PR), however, that solution will not work with module types like for example User or TextChannel (which IIRC was stated in that issue) and I tried it myself, and for some reason it didn't work.

Also, if the name "Option" is very common for a class or something from other libraries, then why not rename those?

**Comment by Angelin01** [comment]
Let me state I regret even proposing solution 1 on that issue. It feels hacky and wrong. I will reinstate the running code on type hints is a bad idea and there's an entire PEP about it.

**Comment by Lulalaby** [comment]
What do we do now?

**Comment by EmreTech** [comment]
After looking at the comment from @Angelin01 on the original issue, PEP, and other places, I have decided that solution 2 would actually be the better method to use. Because of that, I will be reworking this PR to instead change Option to be a Generic type and move the type hinting stuff from the constructor to the `slash_command` decorator.

**Comment by EmreTech** [comment]
Actually, converting `Option` to Generic would break a lot of the library according to @Dorukyum, so instead, I'll change how Options are set to be like this:
```python
@slash_command()
async def hello(ctx, name: str = Option()):
    await ctx.respond(f"Hello, {name}!")
```

**Comment by Dorukyum** [comment]
Further discussion about new style/syntax should be done in the Discord server

**Comment by vcokltfre** [comment]
So instead of discussing technical details on GitHub, the platform where the code is hosted, reviews can be made, etc. with a clear log of all conversation that doesn't get lost in a random channel, it's preferable to do it in a Discord server with none of these advantages?

**Comment by EmreTech** [comment]
> So instead of discussing technical details on GitHub, the platform where the code is hosted, reviews can be made, etc. with a clear log of all conversation that doesn't get lost in a random channel, it's preferable to do it in a Discord server with none of these advantages?

All of the discussions relating to this PR will be in one channel and one channel only. Plus, Dorukyum mentioned the new style/syntax, not how the code would actually work.

**Comment by Dorukyum** [comment]
> So instead of discussing technical details on GitHub, the platform where the code is hosted, reviews can be made, etc. with a clear log of all conversation that doesn't get lost in a random channel, it's preferable to do it in a Discord server with none of these advantages?

The topic got out of the point of this pr, plus Discord will allow more people to share their opinions easily.

**Comment by BobDotCom** [comment]
Merge conflicts here, please resolve

**Comment by Lulalaby** [comment]
Done

**Comment by plun1331** [comment]
Why was support for passing `input_type` to `Option` removed?

**Comment by EmreTech** [comment]
> Why was support for passing `input_type` to `Option` removed?

I thought it wasn't needed. I could re-add `input_type` in the Option `__init__` function if you want me to.

File: examples/app_commands/slash_options.py
Comment (krittick) [inline examples/app_commands/slash_options.py:25]:
```suggestion
    channel: discord.SlashCommandOptionType.channel = Option(description="Select a channel"),
```

No need to go through enums to access this since it's imported via wildcard in `__init__.py`

**Comment by Lulalaby** [comment]
Holy fuck, this got big.
Gonna review over the day.

**Comment by Lulalaby** [comment]
Closing in favor of #1251.
Core team decision.

**Comment by EmreTech** [comment]
I'll do some code fixes and try to improve this PR

**Comment by Lulalaby** [comment]
We reviewed this pr internally and its too messy rn and wasnt in progress for a week / longer.
Clone will complete this within the next days. 

Sorry, but this is the end decision for now.

**Comment by Lulalaby** [comment]
2 months is just too long

**Pull request closed without merging**
